box_error.go

```diff
@@ -0,0 +1,274 @@
+	// "object_name", "access_type".
+	Fields map[interface{}]interface{}
+	// Prev is the previous error in stack.
+	Prev *BoxError
+}
+
+// Error converts a BoxError to a string.
+func (boxError *BoxError) Error() string {
+	return fmt.Sprintf("%s (%s, code 0x%x), see %s:%d",
+		boxError.Msg, boxError.Type, boxError.Code,
+		boxError.File, boxError.Line)
+}
+
+// Depth computes the count of errors in stack, including the current one.
+func (boxError *BoxError) Depth() (depth int) {
+	cur := boxError
+
+	depth = 0
+
+	for cur != nil {
+		cur = cur.Prev
+		depth++
+	}
+
+	return depth
+}
+
+func decodeBoxError(d *decoder) (*BoxError, error) {
+	var l, larr, l1, l2 int
+	var errorStack []BoxError
+	var err error
+	var mapk, mapv interface{}
+
+	if l, err = d.DecodeMapLen(); err != nil {
+		return nil, err
+	}
+
+	for ; l > 0; l-- {
+		var cd int
+		if cd, err = d.DecodeInt(); err != nil {
+			return nil, err
+		}
+		switch cd {
+		case KeyErrorStack:
+			if larr, err = d.DecodeArrayLen(); err != nil {
+				return nil, err
+			}
+
+			errorStack = make([]BoxError, larr)
+
+			for i := 0; i < larr; i++ {
+				if l1, err = d.DecodeMapLen(); err != nil {
+					return nil, err
+				}
+
+				for ; l1 > 0; l1-- {
+					var cd1 int
+					if cd1, err = d.DecodeInt(); err != nil {
+						return nil, err
+					}
+					switch cd1 {
+					case KeyErrorType:
+						if errorStack[i].Type, err = d.DecodeString(); err != nil {
+							return nil, err
+						}
+					case KeyErrorFile:
+						if errorStack[i].File, err = d.DecodeString(); err != nil {
+							return nil, err
+						}
+					case KeyErrorLine:
+						if errorStack[i].Line, err = d.DecodeUint32(); err != nil {
+							return nil, err
+						}
+					case KeyErrorMessage:
+						if errorStack[i].Msg, err = d.DecodeString(); err != nil {
+							return nil, err
+						}
+					case KeyErrorErrno:
+						if errorStack[i].Errno, err = d.DecodeUint32(); err != nil {
+							return nil, err
+						}
+					case KeyErrorErrcode:
+						if errorStack[i].Code, err = d.DecodeUint32(); err != nil {
+							return nil, err
+						}
+					case KeyErrorFields:
+						errorStack[i].Fields = make(map[interface{}]interface{})
+						if l2, err = d.DecodeMapLen(); err != nil {
+							return nil, err
+						}
+						for ; l2 > 0; l2-- {
+							if mapk, err = d.DecodeInterface(); err != nil {
+								return nil, err
+							}
+							if mapv, err = d.DecodeInterface(); err != nil {
+								return nil, err
+							}
+							errorStack[i].Fields[mapk] = mapv
+						}
+					default:
+						if err = d.Skip(); err != nil {
+							return nil, err
+						}
+					}
+				}
+
+				if i > 0 {
+					errorStack[i-1].Prev = &errorStack[i]
+				}
+			}
+		default:
+			if err = d.Skip(); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if len(errorStack) > 0 {
+		return &errorStack[0], nil
+	}
+
+	return nil, nil
+}
+
+func encodeBoxError(enc *encoder, boxError *BoxError) (err error) {
```

Comment:
This is not a very good practice, although unfortunately it occurs in our code:

```suggestion
func encodeBoxError(enc *encoder, boxError *BoxError) error {
```